test-case/check_xrun_injection.sh

```diff
@@ -0,0 +1,156 @@
+
+OPT_OPT_lst['d']='duration' OPT_DESC_lst['d']='aplay/arecord duration in second'
+OPT_PARM_lst['d']=1         OPT_VALUE_lst['d']=10
+
+OPT_OPT_lst['i']='interval'     OPT_DESC_lst['i']='interval time of xrun injection'
+OPT_PARM_lst['i']=1         OPT_VALUE_lst['i']=0.5
+
+OPT_OPT_lst['p']='playback-file'   OPT_DESC_lst['p']='input file for playback'
+OPT_PARM_lst['p']=1         OPT_VALUE_lst['p']=''
+
+OPT_OPT_lst['c']='capture-file'   OPT_DESC_lst['c']='input file for playback'
+OPT_PARM_lst['c']=1         OPT_VALUE_lst['c']=''
+
+OPT_OPT_lst['o']='output'   OPT_DESC_lst['o']='output file dir for capture'
+OPT_PARM_lst['o']=1         OPT_VALUE_lst['o']="$LOG_ROOT/wavs"
+
+OPT_OPT_lst['s']='sof-logger'   OPT_DESC_lst['s']="Open sof-logger trace the data will store at $LOG_ROOT"
+OPT_PARM_lst['s']=0             OPT_VALUE_lst['s']=1
+
+func_opt_parse_option $*
+
+tplg=${OPT_VALUE_lst['t']}
+duration=${OPT_VALUE_lst['d']}
+interval=${OPT_VALUE_lst['i']}
+playback_file=${OPT_VALUE_lst['p']}
+capture_file=${OPT_VALUE_lst['c']}
+output_dir=${OPT_VALUE_lst['o']}
+
+[[ ${OPT_VALUE_lst['s']} -eq 1 ]] && func_lib_start_log_collect
+
+func_lib_setup_kernel_last_line
+func_lib_check_sudo
+func_pipeline_export $tplg "type:playback,capture,both"
+
+xrun_injection()
+{
+    count=1
+    while(true)
+    do
+        ps -ef |grep "$pid" |grep -v grep
+        if [ $? -eq 0 ]; then
+            dlogc "XRUN injection: $count"
+            sudo bash -c "'echo 1 > /proc/asound/card0/$pcm/sub0/xrun_injection'"
+            sleep $interval
+            let count++
+	else
+            break # aplay/arecord is finished, stop xrun injection
+        fi
+    done
+}
+
+run_test()
+{
+    dlogc "$cmd -D$dev -r $rate -c $channel -f $fmt -d $duration $file -q"
+    $cmd -D$dev -r $rate -c $channel -f $fmt -d $duration $file -q &
+    pid=$!
+    # do xrun injection
+    xrun_injection
+    # check aplay/arecord return value
+    wait $pid
+    if [ $? != 0 ]; then
+        dloge "$pipeline_type on $pcm failed."
+        exit 1
+    fi
+}
+
+get_audio_file()
+{
+    # for playback
+    if [[ $pipeline_type == "playback" ]]; then
+        if [[ -z $playback_file ]]; then
+            dlogi "no playback file specified, use /dev/zero as dummy playback source"
+            file=/dev/zero
+        elif [[ ! -f $playback_file ]]; then
+            dlogw "$playback_file does not exist, use /dev/zero as dummy playback source"
+            file=/dev/zero
+        else
+	    file=$playback_file
+            dlogi "using $playback_file as playback source"
+	fi
+    fi
+
+    # for capture
+    if [[ $pipeline_type == "capture" ]]; then
+        if [[ -z $capture_file ]]; then
+            dlogi "no capture file specified, use /dev/null as dummy capture output"
+            file=/dev/null
+        else
+            mkdir -p $output_dir
+            file=$output_dir/${capture_file}_${dev}.wav
+            dlogi "using $file as capture output"
+	fi
+    fi
+}
```

Comment:
1.  line 112 can be move to the top, because this folder just need to create once
2.  after remove the line 112, you can find the capture & playback use the same logic

I guest use `/dev/zero` & `/dev/null` as the default value of `$playback_file` & `$capture_file`  
then here you just need to check target file whether exist, if not exist, you can direct exit as 2/1

from your code logic, I think this maybe help for your code more clear  
https://github.com/thesofproject/sof-test/blob/master/test-case/multiple-pipeline-capture.sh#L53  
https://github.com/thesofproject/sof-test/blob/master/test-case/multiple-pipeline-capture.sh#L84  

here use the associative array about bash, you need to `declare -A ` for your array

Reply:
Test script is updated. Good example.